src/qibolab/platform.py

```diff
@@ -167,14 +167,14 @@ def disconnect(self):
                 instrument.disconnect()
         self.is_connected = False
 
-    def _execute(self, sequence, channel_cfg, options, **kwargs):
+    def _execute(self, sequences, channel_cfg, options, **kwargs):
```

Comment:
This change is independent from the unification effort. This is because I have removed the unrolling. Now after splitting into batches platform submits each batch as a list to the instrument, instead of unrolling them into a single sequence. If particular instrument needs unrolling, the function is still there, they can call the unrolling function themselves. This change was introduced to simplify the ZI driver, which was previously trying very hard to un-unroll the sequences back.

Reply:
This is completely ignoring the idea of unrolling, which is controlled by the Qibolab user (within all the limits of the device used), not by the driver developer.

Please, show an example for the ZI driver, and we can improve on that side, instead of removing the feature for everyone else.

Since instruments should be capable of executing arbitrary sequences (within their memory limitations) unrolled sequences should be perfectly legitimate and indistinguishable from similar sequences requested by the user. If ZI is not capable of executing them, then we have a problem with the ZI driver.

Reply:
I finally managed to find this discussion!

To be fair, the current implementation of unrolling is quite useless for QM as well, due to various issues previously discussed in other places (not being able to bypass the compiler, etc.). Maybe "useless" is strong word, it could still speed up a few routines, but most likely far from optimal.

Regardless, I do not see any issue with the change implemented here. The batching, which is a bit more complicated code-wise, is still done at the platform level. The only overhead is calling `unroll_sequences` in every driver, but this is just a single line of code. Having the list of sequences in the driver is even more flexible if the driver developer does not want to naively use `unroll_sequences` for some reason, for example if the underlying instrument provides a better way to unroll.

I would certainly not block the PR for this. It is also not interface facing, it can be easily changed in the future if needed without affecting other repos and users.

Reply:
> This is completely ignoring the idea of unrolling, which is controlled by the Qibolab user

How is it controlled by user? As far as I know, qibocal does the following: 1. if user specified the unrolling option, it uses the `execute_pulse_sequences` method, to submit the list of pulse sequences, 2. if unrolling option is not specified, qibocal uses either `sweep` or `execute_pulse_sequence` method to submit a single sequence. From user's perspective they are either submitting a list of sequences, or a single sequence. After the unification of all these execution functions, there is one `execute` method only that accepts a list of sequences. If user wants to submit their sequences one by one, they can submit lists of sequences of length one, if they want to submit all, they submit all. They only know, that submitting all may bring some performance benefits, but how exactly, does not matter, it is internal implementation detail. With my change, from the perspective of users experience changes exactly nothing.

I believe some details are misunderstood about this change. By removing the call to the unrolling function before execution is submitted to instrument, does not mean performance optimization brought by unrolling are gone for the zi driver. It still executes the entire batch as one, but it is free from the structure enforced by the unrolling in platform level. You could say "the driver does unrolling its own way". 

> Calling unroll_sequences in all drivers is not just overhead, but a design flow. Drivers should be able to play arbitrary sequences (bound to hardware limitations), and they should not distinguish between an unrolled one and not unrolled one.

Sure, each driver should be able to execute arbitrary sequence. I don't see how moving the place where `unroll_sequences` is called (or not called) changes this. In case of zi driver specifically, there is a possibility to do more optimized and precise execution if you know that you are executing list of independent experiments as compared to a single big experiment, but you are still able to correctly execute single big experiment as well for sure.

> Requiring the drivers to do more makes it heavier and more difficult to maintain many drivers

I mean, this claim taken out of context and in general is true. If you can reliably generalize some operations into platform, or wherever, instead of repeating in each driver, you should certainly do it. However it has negligible overlap with the discussion in this thread, which is just about moving the place of a function call.

Reply:
> Sure, each driver should be able to execute arbitrary sequence. I don't see how moving the place where `unroll_sequences` is called (or not called) changes this. In case of zi driver specifically, there is a possibility to do more optimized and precise execution if you know that you are executing list of independent experiments as compared to a single big experiment, but you are still able to correctly execute single big experiment as well for sure.

Let's focus on this specific point: why independent experiments are better supported than a long one?